### PR TITLE
mpc85xx: add support for Aerohive BR200-WP

### PR DESCRIPTION
**Pull request by CHKDSK88**
The following adds the Aerohive BR200-WP router to OpenWrt under
the mpc85xx/p1010 subtarget.

Hardware:
- SoC: Freescale P1011
- NOR: Intel JS28F512M29EWH 64MB
- Memory: 2x Nanya NT5TU64M16GG-AC 128MB (Total of 256MB)
- 2.4GHz WiFi: Atheros AR9390-AL1A
- Eth1: Atheros AR8035-A PoE
- 2x LED
- 1x Button
- PoE PSE

Flashing:
1. Hook into UART (9600 baud) and enter U-Boot. You may need to enter a
password of administrator or AhNf?d@ta06 if prompted.
2. Once in U-Boot, tftp boot the initramfs image:
   ```
   dhcp; setenv serverip 192.168.1.3;
   tftpboot 0x2004000 openwrt-mpc85xx-p1010-aerohive_br200-wp-initramfs-kernel.bin;
   bootm 0x2004000;
   ```
3. Once booted, scp over the sysupgrade file and sysupgrade the device
to flash LEDE to the NOR.

Note:

MAC assigns are taken from stock firmware:
```
Name        MAC addr      Mode       State Chan(Width) VLAN   Radio      Hive       SSID
-------- -------------- --------     ----- ----------- ---- ---------- ---------- ---------
Mgt0     08ea:44XX:XXc0    -           U     -            1     -        hive0        -
Eth0     08ea:44XX:XXc0 wan            U     -            -     -          -          -
Eth1     08ea:44XX:XXc2 access         D     -            -     -        hive0        -
Eth2     08ea:44XX:XXc3 access         D     -            -     -        hive0        -
Eth3     08ea:44XX:XXc4 access         D     -            -     -        hive0        -
Eth4     08ea:44XX:XXc5 access         D     -            -     -        hive0        -
Wifi0    08ea:44XX:XXd0 access         U     1(20MHz)     -  radio_ng0     -          -
Wifi0.1  08ea:44XX:XXd4 access         D     1(20MHz)     -  radio_ng0   hive0        -
```
Note2:
PoE PSE could be managed with `realtek-poe` package. Example port
config:
```
config port
        option enable   '1'
        option id       '4'
        option name     'lan2'
        option poe_plus '0'
        option priority '2'
config port
        option enable   '1'
        option id       '3'
        option name     'lan1'
        option poe_plus '0'
        option priority '1'
```

### CODE REVIEW

File: target/linux/mpc85xx/files/arch/powerpc/boot/dts/br200-wp.dts [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+					port@6 {
+						reg = <6>;
+						ethernet = <&enet0>;
+						phy-mode = "rgmii-id";
+
+						fixed-link {
+							speed = <1000>;
+							full-duplex;
+						};
+					};
```
Comment (Ansuel) [inline target/linux/mpc85xx/files/arch/powerpc/boot/dts/br200-wp.dts:219]: port 0 not connected?

**Comment by Hurricos** [comment]
Hey @CHKDSK88,

I'll test tomorrow when I'm back at Laboratory B. I hadn't noticed there was a U-ART.

This board has other neighbors, in particular a 4G variant; have you seen it?

File: target/linux/mpc85xx/image/p1010.mk [hunk 2]
```diff
@@ -14,6 +14,26 @@ define Build/spi-loader-okli
...
+  BLOCKSIZE := 128k
+  KERNEL_NAME := simpleImage.br200-wp
+  KERNEL := kernel-bin | uImage none
+  KERNEL_INITRAMFS := kernel-bin | uImage none
+  KERNEL_ENTRY := 0x1000000
+  KERNEL_LOADADDR := 0x1000000
+  KERNEL_SIZE := 8m
+  IMAGES := fdt.bin sysupgrade.bin
+  IMAGE/fdt.bin := append-dtb
+  IMAGE/sysupgrade.bin := append-dtb | pad-to 256k | check-size 256k | \
```
Comment (Hurricos) [inline target/linux/mpc85xx/image/p1010.mk:29]: Seems like, as with other mpc85xx Aerohives, you've found a hardcoded bootcmd; with an 8m limit on kernel-size, and using an uncompressed kernel, won't we be running out of space fairly soon?

**Comment by Hurricos** [comment]
> NAND: Intel JS28F512M29EWH 64MB

This is NOR, not NAND: https://pdf1.alldatasheet.com/datasheet-pdf/view/504370/MICRON/JS28F512M29EWHA.html

**Comment by Hurricos** [comment]
@CHKDSK88 Surprising, this variant of the BR200 [really does support acting as POE PSE](https://docs.aerohive.com/330000/docs/help/english/6.8r3/hm/full/Content/hw/br/BR200.htm).

I don't see any obvious PSE chips on the [FCC filings for the BR200-WP](https://fccid.io/WBV-BR200-WP/Internal-Photos/Internal-Photos-1829577.pdf); could you upload photos of your board?

File: target/linux/mpc85xx/image/p1010.mk (hunk 2 above)
Comment (CHKDSK88) [reply target/linux/mpc85xx/image/p1010.mk:29]: I use simpleImage with LZMA self-extractor:
https://docs.kernel.org/powerpc/bootwrapper.html

Now kernel have 3MB only:
```

U-Boot 2009.11 (Sep 17 2015 - 00:35:03)

CPU:   P1011E, Version: 1.1, (0x80ed0011)
Core:  E500, Version: 5.1, (0x80212051)
Clock Configuration:
       CPU0:660  MHz,
       CCB:330  MHz,
       DDR:330  MHz (660 MT/s data rate) (Asynchronous), LBC:20.625 MHz
L1:    D-cache 32 kB enabled
       I-cache 32 kB enabled
I2C:   ready
SPI:   ready
DRAM:  Configuring DDR for 660 MT/s data rate
DDR: 256 MB
FLASH: 64 MB
L2:    256 KB enabled
MMC:

    PCIE2: disabled

    PCIE1 connected to Slot 2 as Root Complex (base addr ffe0a000)
               Scanning PCI bus 01
        01  00  168c  0030  0280  ff
    PCIE1 on bus 00 - 01

In:    serial
Out:   serial
Err:   serial
Net:   initialization for Atheros AR8327/AR8328,addr 4
AR8327/AR8328 v1.1 is found!
set for PHY4
initialization for Atheros AR8327/AR8328,addr 0
eth0, eth1
Hit any key to stop autoboot:  0
Hit any key to interrupt boot from flash:  0
## Booting kernel from Legacy Image at ee840000 ...
   Image Name:   POWERPC OpenWrt Linux-5.15.106
   Created:      2023-04-12  11:21:21 UTC
   Image Type:   PowerPC Linux Kernel Image (uncompressed)
   Data Size:    3094436 Bytes =  3 MB
   Load Address: 01000000
   Entry Point:  01000000
   Verifying Checksum ... OK
## Loading init Ramdisk from Legacy Image at 02000000 ...
   Image Name:   OpenWrt fake ramdisk
   Created:      2023-04-12  11:21:21 UTC
   Image Type:   PowerPC Linux RAMDisk Image (uncompressed)
   Data Size:    0 Bytes =  0 kB
   Load Address: 00000000
   Entry Point:  00000000
   Verifying Checksum ... OK
## Flattened Device Tree blob at ec000000
   Booting using the fdt blob at 0xec000000
   Loading Kernel Image ... OK
OK
   Loading Ramdisk to 10000000, end 10000000 ... OK
   Loading Device Tree to 00ff9000, end 00fff178 ... OK
ft_fixup_l2cache: FDT_ERR_NOTFOUND

zImage starting: loaded at 0x01000000 (sp: 0x012f4fa0)
Allocating 0x91626c bytes for kernel...
Decompressing (0x00000000 <- 0x01014000:0x012f37a4)...
Done! Decompressed 0x8dd7fc bytes

Linux/PowerPC load: console=ttyS0,9600
Finalizing device tree... flat tree at 0x12f5cc0
[    0.000000] Memory CAM mapping: 256 Mb, residual: 0Mb
[    0.000000] Linux version 5.15.106 (krolik@ubuntu) (powerpc-openwrt-linux-musl-gcc (OpenWrt GCC 12.2.0 r21408+1090-7dbac3433fef) 12.2.0, GNU ld (GNU Binutils) 2.40.0) #0 Wed Apr 12 11:21:21 2023
```

File: target/linux/mpc85xx/files/arch/powerpc/boot/dts/br200-wp.dts (hunk 1 above)
Comment (CHKDSK88) [reply target/linux/mpc85xx/files/arch/powerpc/boot/dts/br200-wp.dts:219]: Yep. Only port 6 is connected to eth0.

File: target/linux/mpc85xx/image/p1010.mk (hunk 2 above)
Comment (Hurricos) [reply target/linux/mpc85xx/image/p1010.mk:29]: GAH! How did we not notice this when we were fighting with the [Aerohive HiveAP 330](https://github.com/openwrt/openwrt/pull/4897)?

It must have just been who was around at the time. simpleImage comes with its own challenges, clearly they've been figured out for the host of mpc85xx boards OpenWrt supports.

**Comment by Hurricos** [comment]
[Here](https://paste.c-net.org/TilesServants) is a copy of `drivers/ah_drivers/ah_pse.c`, which is the source for the `pse` command, as extracted from the u-boot archive in my previous post.

**Comment by Hurricos** [comment]
To translate to what it means for you, @CHKDSK88: we would interact with this PSE chip in userspace using `realtek-poe`; we just need to figure out whether the chip is connected through i2c or serial. I'm guessing serial; we should work out the baud rate it wants and on which port, and then start talking to the chip, as is done in [`realtek-poe`](https://github.com/Hurricos/realtek-poe) source.

**Comment by CHKDSK88** [comment]
I repaired my PoE. Common Mode inductor was burned in both ports. It works well with realtek-poe package:

```
root@OpenWrt:/# cat /etc/config/poe
config global
        option budget   '170'

config port
        option enable   '1'
        option id       '4'
        option name     'lan2'
        option poe_plus '0'
        option priority '2'
config port
        option enable   '1'
        option id       '3'
        option name     'lan1'
        option poe_plus '0'
        option priority '1'
root@OpenWrt:/# ubus call poe info
{
        "firmware": "v64.18",
        "mcu": "ST Micro ST32F100 Microcontroller",
        "budget": 170.000000,
        "consumption": 2.100000,
        "ports": {
                "lan1": {
                        "priority": 1,
                        "mode": "PoE",
                        "status": "Searching"
                },
                "lan2": {
                        "priority": 2,
                        "mode": "PoE",
                        "status": "Delivering power",
                        "consumption": 2.100000
                }
        }
}
```

**Comment by CHKDSK88** [comment]
Thank you @Hurricos.

**Comment by Hurricos** [comment]
@CHKDSK88: It's utterly brilliant that realtek-poe works out-of-the-box on this hardware. I'm very excited to see it :^)

**Comment by chunkeey** [comment]
merged. Thanks!

**Comment by svanheule** [comment]
> based on the 12-byte frame format, it looks very certainly like this _is_ the same code. @mrnuke @svanheule you may be interested in some vendor source-code for interacting with the Broadcom PSE equipment -- esp. code to reflash the PSE chips themselves, in `ah_pse_do_load_image` -- though in this case the frames are not proxied through a microcontroller.

From the firmware reverse engineering I did, I had the impression that the MCU protocol closely mimics the PSE controller protocol. Perhaps this was (initially) done to have some sort of drop-in proxy to perform real-time monitoring, but that I can only guess.

**Comment by neheb** [comment]
@CHKDSK88 why is the ath9k node a gpio controller?

**Comment by CHKDSK88** [comment]
> @CHKDSK88 why is the ath9k node a gpio controller?

Copy-paste leftover.

**Pull request closed without merging**